Android/app/src/main/java/app/intra/net/socks/GoIntraListener.java

```diff
@@ -0,0 +1,93 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package app.intra.net.socks;
+
+import android.content.Context;
+import android.os.Bundle;
+import app.intra.sys.Names;
+import com.google.firebase.analytics.FirebaseAnalytics;
+import intra.RetryStats;
+import intra.TCPSocketSummary;
+import intra.UDPSocketSummary;
+
+/**
+ * This is a callback class that is passed to our go-tun2socks code.  Go calls this class's methods
+ * when a socket has concluded, with performance metrics for that socket, and this class forwards
+ * those metrics to Firebase.
+ */
+public class GoIntraListener implements tunnel.IntraListener {
+
+  // UDP is often used for one-off messages and pings.  The relative overhead of reporting metrics
+  // on these short messages would be large, so we only report metrics on sockets that transfer at
+  // least this many bytes.
+  private static final int UDP_THRESHOLD_BYTES = 10000;
+
+  private final Context context;
+  GoIntraListener(Context context) {
+    this.context = context;
+  }
+
+  @Override
+  public void onTCPSocketClosed(TCPSocketSummary summary) {
+    // We had a functional socket long enough to record statistics.
+    // Report the BYTES event :
+    // - UPLOAD: total bytes uploaded over the lifetime of a socket
+    // - DOWNLOAD: total bytes downloaded
+    // - PORT: TCP port number (i.e. protocol type)
+    // - TCP_HANDSHAKE_MS: TCP handshake latency in milliseconds
+    // - DURATION: socket lifetime in seconds
+    // - TODO: FIRST_BYTE_MS: Time between socket open and first byte from server, in milliseconds.
+
+    Bundle bytesEvent = new Bundle();
+    bytesEvent.putLong(Names.UPLOAD.name(), summary.getUploadBytes());
+    bytesEvent.putLong(Names.DOWNLOAD.name(), summary.getDownloadBytes());
+    bytesEvent.putInt(Names.PORT.name(), summary.getServerPort());
+    bytesEvent.putInt(Names.TCP_HANDSHAKE_MS.name(), summary.getSynack());
+    bytesEvent.putInt(Names.DURATION.name(), summary.getDuration());
+    FirebaseAnalytics.getInstance(context).logEvent(Names.BYTES.name(), bytesEvent);
+
+    RetryStats retry = summary.getRetry();
+    if (retry != null) {
+      // Prepare an EARLY_RESET event to collect metrics on success rates for splitting:
+      // - BYTES : Amount uploaded before reset
+      // - CHUNKS : Number of upload writes before reset
+      // - TIMEOUT : Whether the initial connection failed with a timeout.
+      // - SPLIT : Number of bytes included in the first retry segment
+      // - RETRY : 1 if retry succeeded, otherwise 0
+      Bundle resetEvent = new Bundle();
+      resetEvent.putInt(Names.BYTES.name(), retry.getBytes());
+      resetEvent.putInt(Names.CHUNKS.name(), retry.getChunks());
+      resetEvent.putInt(Names.TIMEOUT.name(), retry.getTimeout() ? 1 : 0);
```

Comment:
Is there an advantage in using `putInt` vs `putBoolean`?

Reply:
The immediate reason is that these attributes are already in use, so it's probably a bad idea to change their types now.  The original reason is that Firebase only knows about numeric and string-valued attributes, and places a limit on the number of string-valued attributes per project.